### DISCUSSION TITLE
Trouble Disabling AES-NI via Environment Variable in C Code.

### DISCUSSION THREAD

**Discussion by aahajj** [openssl/openssl#22313] [Q&A]
I'm trying to disable AES-NI in OpenSSL by setting the `OPENSSL_ia32cap` environment variable within my C code as shown below:

```c
#define NO_AESNI
int main() {
    #ifdef NO_AESNI 
    const char *varName = "OPENSSL_ia32cap";
    #ifdef NO_SSSE3
    const char *varValue = "~0x200020200000000";
    #else
    const char *varValue = "~0x200000200000000";
    #endif
    if (setenv(varName, varValue, 1) != 0) {
        perror("Error setting environment variable");
        return 1;
    }

    char *retrievedValue = getenv(varName);
    if (retrievedValue != NULL) {
        printf("Environment variable %s has value: %s\n", varName, retrievedValue);
    } else {
        printf("Environment variable %s not found\n", varName);
    }
    #endif
    // Rest of my code
}
```
Here is an example of running the code with AES-NI:
```text
[OpenSSL] running AES-128-ECB benchmark...
[OpenSSL] 3.482924 seconds for 10000000 iterations, 4096 bytes message
[OpenSSL] running AES-128-CTR benchmark...
[OpenSSL] 3.877055 seconds for 10000000 iterations, 4096 bytes message
```
And this is the output after setting the Environment Variable:

```text
Environment variable OPENSSL_ia32cap has value: ~0x200000200000000
[OpenSSL] running AES-128-ECB benchmark...
[OpenSSL] 3.485387 seconds for 10000000 iterations, 4096 bytes message
[OpenSSL] running AES-128-CTR benchmark...
[OpenSSL] 3.893353 seconds for 10000000 iterations, 4096 bytes message
```

As you can see, it doesn't seem to work, and AES-NI is not disabled. I've noticed that when I set the environment variable in the terminal and run `openssl-speed`, it does work. Can anyone provide insights into why this is not working when set within the code and suggest a solution to disable AES-NI programmatically within the code?

Edit: typo.

**Comment by paulidale** [comment]
I'd guess that the environment variable is being read before you set it.
The code only reads it once.

As to when or why, I'm not sure.

**Comment by aahajj** [comment]
Thanks for the reply. I tried setting the environment variable in the Terminal before running my code by running`export OPENSSL_ia32cap="~0x200000200000000"
` and it seems to work. This time the result was much better:
`[OpenSSL] 187.565601 seconds for 10000000 iterations, 4096 bytes message`.
